gravshort-tree.c

```diff
@@ -248,26 +249,37 @@ int force_treeev_shortrange(TreeWalkQueryGravShort * input,
                     }
                 }
 
-                h = All.ForceSoftening[ptype];
-                otherh = All.ForceSoftening[nop->f.MaxSofteningType];
+                ph = All.ForceSoftening[ptype];
+                nodeh = All.ForceSoftening[nop->f.MaxSofteningType];
                 if(All.AdaptiveGravsoftForGas) {
                     if(ptype == 0)
-                        h = input->Soft;
-                    if(otherh < nop->u.d.hmax)
-                        otherh = nop->u.d.hmax;
+                        ph = input->Soft;
+                    if(nodeh < nop->u.d.hmax)
+                        nodeh = nop->u.d.hmax;
                 }
-                if(h < otherh)
-                {
-                    h = otherh;
-                    if(r2 < h * h)
-                    {
-                        if(nop->f.MixedSofteningsInNode)
-                        {
+                /* We want to open the node when it is possible
+                 * that some but not all of it is within
+                 * the softening length; opening is the only way
+                 * we can tell how much.*/
+                /* This only happens if at least some of the 
+                 * node is within the softening length.*/
+                h = DMAX(ph, nodeh);
+                if(r2 < (h + nop->len/2)*(h + nop->len/2)) {
+                    /* a) the node softening dominates, but the node
+                     * particles do not all have the same softening value,
+                     */
+                    if(ph < nodeh) {
+                        if(nop->f.MixedSofteningsInNode) {
                             no = nop->u.d.nextnode;
-
                             continue;
                         }
                     }
+                    /* b) No need to open if all the node is within
+                     * the softening length: all is softened.*/
+                    if(r2 > (h - nop->len/2)*(h - nop->len/2)) {
```

Comment:
What happens if h < hop->len/2?

Reply:
In practice this becomes a less stringent version of the "check whether we are within the node" but that is kind of accidental and needs to be cleaned up.